src/ray.rs

```diff
@@ -211,14 +215,83 @@ impl Ray {
 
         tmax >= tmin && tmax >= 0.0
     }
+
+    /// Implementation of the [Möller-Trumbore triangle/ray intersection algorithm]
+    /// (https://en.wikipedia.org/wiki/M%C3%B6ller%E2%80%93Trumbore_intersection_algorithm).
+    /// Returns the distance to the intersection, as well as
+    /// the u and v coordinates of the intersection.
+    /// The distance is set to +INFINITY if the ray does not intersect the triangle, or hits
+    /// it from behind.
+    pub fn intersects_triangle(&self,
+                               a: &Point3<f32>,
+                               b: &Point3<f32>,
+                               c: &Point3<f32>)
+                               -> (f32, f32, f32) {
+        const EPSILON: f32 = 0.00001;
+
+        let a_to_b = *b - *a;
+        let a_to_c = *c - *a;
+
+        // Begin calculating determinant - also used to calculate u parameter
+        // u_vec lies in view plane
+        // |u_vec| = |a_to_c|*sin(a_to_c, dir) := length of a_to_c in view_plane
+        let u_vec = self.direction.cross(&a_to_c);
+
+        // If determinant is near zero, ray lies in plane of triangle
+        // The determinant corresponds to the parallelepiped volume:
+        // det = 0 => [dir, a_to_b, a_to_c] not linearly independant
+        let det = a_to_b.dot(&u_vec);
+
+        // Only testing positive bound, thus enabling backface culling
+        // If backface culling is not desired write:
+        // det < 0.00001 && det > -0.00001
+        if det < EPSILON {
+            return (INFINITY, 0.0, 0.0);
+        }
+
+        let inv_det = 1.0 / det;
+
+        // Vector from point a to ray origin
+        let a_to_origin = self.origin - *a;
+
+        // Calculate u parameter
+        let u = a_to_origin.dot(&u_vec) * inv_det;
+
+        // Test bounds: u < 0 || u > 1 => outside of triangle
+        if u < 0.0 || u > 1.0 {
+            return (INFINITY, u, 0.0);
+        }
+
+        // Prepare to test v parameter
+        let v_vec = a_to_origin.cross(&a_to_b);
+
+        // Calculate v parameter and test bound
+        let v = self.direction.dot(&v_vec) * inv_det;
+        // The intersection lies outside of the triangle
+        if v < 0.0 || u + v > 1.0 {
+            return (INFINITY, u, v);
+        }
+
+        let dist = a_to_c.dot(&v_vec) * inv_det;
+
+        if dist > EPSILON {
+            (dist, u, v)
+        } else {
+            (INFINITY, u, v)
+        }
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use ray::Ray;
     use aabb::AABB;
-    use nalgebra::{Point3, Vector3};
+    use nalgebra::{Point3, Vector3, Cross, Dot};
     use rand::{Rng, StdRng, SeedableRng};
+    use std::f32::INFINITY;
+    use std::cmp;
+
+    const EPSILON: f32 = 0.00001;
```

Comment:
We define this so often there should be either a global constant or it should be configurable everywhere.